Step2 - 2단계 문자열 덧셈계산기 리뷰요청 드립니다.

aksgudwns opened the pull request.

aksgudwns committed:
fix - setTest 수정사항 반영

assertThat을 통한 체크로 수정

aksgudwns committed:
doc - 문자열 덧셈 계산기 요구사항 추가

문자열 덧셈 계산기 요구사항 추가

aksgudwns committed:
feat - 문자열계산기 빈문자null체크, 숫자 유효성체크 기능 작성

- 빈문자null체크 기능 작성
- 입력된 문자열이 숫자인지 체크하는 유틸로직 작성

aksgudwns committed:
feat - 구현기능 테스트, splitor만 입력했을 때 오류 추가

- 기본기능 테스트
- splitor만 입력했을 경우 오류나도록 추가

aksgudwns committed:
feat - custom delimiter 기능 구현

- custom delimiter로 구분하여 합을 구할 수 있는 기능 추가
- custom delimiter와 default delimiter를 혼합하였을 경우 테스트케이스 추가

dhmin5693 commented on src/main/java/string_add_calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+    private final String splitter = ",|:";
+
+    public int splitAndSum(String text) {
+        if(text == null || text.isEmpty()) return 0;
```
자바에서는 한 줄짜리에서도 중괄호를 사용하는 컨벤션이 있습니다.
https://naver.github.io/hackday-conventions-java/#need-braces

aksgudwns replied:
수정하겠습니다! 밑에 RuntimeException리턴하는 부분도 같이 수정하였습니다!

dhmin5693 commented on src/main/java/string_add_calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+
+    private String[] splitText(String text) {
+        String[] numbers = text.split(splitter);
+        Matcher m = Pattern.compile("//(.)\n(.*)").matcher(text);
```
`Pattern.compile("//(.)\n(.*)")` 은 한 번만 생성되어도 될 것 같은데 splitText가 실행될 때마다 새로운 인스턴스가 생성되겠네요.
자원을 더 아낄 수 있는 방법은 없을까요?

aksgudwns replied:
Pattern.compile("//(.)\n(.*)") 부분을 상수 필드로 선언하고, 
m도 가독성을 위해 matcher로 이름 수정하였습니다!

dhmin5693 commented on src/main/java/string_add_calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class StringAddCalculator {
+
+    private final String splitter = ",|:";
```
이 클래스는 인스턴스로 생성될 필요가 없어보입니다.
생성자 사용을 막는건 어떨까요?

splitter 는 무조건 같은 값을 같는 불변 상태로 만드셨는데요, 상수화하는게 낫지 않을까요?

aksgudwns replied:
private 생성자를 추가하여 인스턴스 생성을 막겠습니다! 
splitter도 상수화 하였습니다!

dhmin5693 commented on src/main/java/string_add_calculator/util/StringAddCalculatorUtil.java:
```diff
@@ -0,0 +1,23 @@
+package string_add_calculator.util;
+
+public class StringAddCalculatorUtil {
```
StringAddCalculator 만을 위한 Util 클래스라면 따로 분리할 필요는 없어보입니다.
util이라는 이름이 붙는 클래스엔 조금 더 공용이라는 느낌이 드는 내용을 묶는건 어떻게 생각하시나요?

보통 util으로 빼는 클래스는 공용의 느낌이 크다보니 다른 코드에서도 의존하는 경우가 많습니다.
그 결과 조금만 수정하더라도 의도치 않게 큰 버그가 나기도 하더라구요.
그래서 util로 빼는 대상은 정말 보수적으로 선정하시길 하는 마음에 의견 남겨보았습니다.

aksgudwns replied:
말씀주신대로 util로 빼는 메서드는 신중히 생각하겠습니다.
util클래스는 삭제하였습니다.

dhmin5693 commented on src/test/java/string_add_calculator/StringAddCalculatorUtilTest.java:
```diff
@@ -0,0 +1,50 @@
+
+    StringAddCalculatorUtil stringAddCalculatorUtil = new StringAddCalculatorUtil();
+
+    @Test
+    void 문자열_하나_유효성_체크() {
+        assertThat(stringAddCalculatorUtil.isCharIntValue('a')).isFalse();
+    }
+
+    @Test
+    void 음수_하나_유효성_체크() {
+        assertThat(stringAddCalculatorUtil.isCharIntValue('-')).isFalse();
+    }
+
+    @Test
+    void 숫자_하나_최소값_유효성_체크() {
+        assertThat(stringAddCalculatorUtil.isCharIntValue('0')).isTrue();
+    }
+
+    @Test
+    void 숫자_하나_최댓값_유효성_체크() {
+        assertThat(stringAddCalculatorUtil.isCharIntValue('9')).isTrue();
+    }
```
검증할 변수가 조금 다를 뿐, 유효성을 체크한다는 점에선 같은 목적을 가지고 있는 메소드입니다.
이럴 때 `@ParameterizedTest` 를 활용해보시면 훨씬 도움이 될거에요.

aksgudwns replied:
말씀주신 부분 반영하였습니다!
같은 목적을 가지고 있는 메소드들은 @ParameterizedTest를 적극 활용하겠습니다.
나머지 테스트 케이스 중에서도 같은목적을 지닌 것으로 보이는 메소드들은 @ParameterizedTest 하나의 메소드로 통합하였습니다.

aksgudwns replied:
테스트 메소드들도 이름에 통일성이 없어서 앞에는 메서드 이름_xxx_테스트로 통일하였습니다.

dhmin5693 commented on src/main/java/string_add_calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,32 @@
+    public int splitAndSum(String text) {
+        if(text == null || text.isEmpty()) return 0;
+        String[] numbers = splitText(text);
+        if(numbers.length == 0) throw new RuntimeException();
```
미션에 `RuntimeException` 예외를 던지라고만 명시되어 있지만 조금 더 구체화해보는건 어떤가요?
예를 들면 메시지를 같이 준다던가? 아니면 다른 예외로 바꿔서 보낸다던가?

aksgudwns replied:
입력값이 잘못된 오류라, IllegalArgumentException을 던지도록 수정해보았습니다.
에러메시지도 추가하였습니다.

dhmin5693 reviewed (changes requested):
안녕하세요!
미션을 정말 잘 완수해주셨네요 👍 

조금만 고치면 정말 좋을 것 같아서 약간의 여지를 남겨보았는데요,
수정 후 다시 요청주시면 빠르게 확인해보도록 하겠습니다!

혹시라도 어려운 점 있으시다면 `#wootecam-pro-qna` 채널이나 DM 보내주세요. 🙇‍♂️

aksgudwns committed:
Refactor - 코드리뷰 반영

- 자바코드컨벤션 한 줄짜리에서도 중괄호 사용하도록 수정
- Pattern.compile("//(.)\n(.*)") 비즈니스로직상 메서드 호출할때마다 인스턴스 생성될 필요 없으므로 상수로 변경
- StringAddCalculator splitter 상수화
- StringAddCalculatorUtil 제거 Util 클래스는 좀 더 보수적으로 생각
- 테스트메서드 같은 목적을 가진 메서드들은 ParameterizedTest로 통합
- RuntimeException IllegalArgumentException과 메시지 추가로 오류내용 구체화

aksgudwns commented:
구체적이고 친절하게 리뷰해주셔서 감사합니다!
StringAddCalculator 과제 내용이 간단하다고 생각하였는데, 간과하고 있는 부분이 생각보다 많았고, 덕분에 많이 배웠습니다.
말씀주신부분 수정반영했습니다. 다시 리뷰부탁드리겠습니다!

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요 😁

수고 많으셨습니다.
이번 과제에서는 충분히 학습이 완료된 것 같으니 다음 미션 진행해주세요. 💪
